File: classifications/outdoor.go
package classifications

import (
	"fmt"
	"sync"

	"github.com/jatkinson1000/archeryutils/handicaps"
	"github.com/jatkinson1000/archeryutils/rounds"
)

// outdoorGroupData stores the classification config for one outdoor category.
type outdoorGroupData struct {
	classes       []string
	classesLong   []string
	classHC       []float64
	minDists      []float64
	maxDists      []float64
	prestigeRounds []string
}

var (
	outdoorOnce sync.Once
	outdoorDict map[string]outdoorGroupData
)

// outdoorRoundsOnce loads the outdoor round set once.
var (
	outdoorRoundsOnce sync.Once
	outdoorRoundsMap  map[string]*rounds.Round
)

func allOutdoorRounds() map[string]*rounds.Round {
	outdoorRoundsOnce.Do(func() {
		outdoorRoundsMap = make(map[string]*rounds.Round)
		for k, v := range rounds.AGBOutdoorImperial() {
			outdoorRoundsMap[k] = v
		}
		for k, v := range rounds.AGBOutdoorMetric() {
			outdoorRoundsMap[k] = v
		}
		for k, v := range rounds.WAOutdoor() {
			outdoorRoundsMap[k] = v
		}
	})
	return outdoorRoundsMap
}

func buildOutdoorDict() map[string]outdoorGroupData {
	outdoorOnce.Do(func() {
		agbAgeData := loadAgeData()
		agbBowstyleData := loadBowstyleData()
		classInfo := loadClassesOut()
		classes := classInfo.Classes
		classesLong := classInfo.ClassesLong
		n := len(classes)

		outdoorDict = make(map[string]outdoorGroupData)

		for _, bowstyle := range OutdoorBowstyles {
			for _, gender := range AllGenders {
				for _, age := range AllAges {
					groupname := GroupName(bowstyle, gender, age)
					ageName := age.String()
					ad := agbAgeData[ageName]
					bd := agbBowstyleData[bowstyle.String()]

					maxDists := maxDistsForGender(ad, gender)
					ageCat := ad.Step

					delta := GetAgeGenderStep(gender, ageCat, bd.AgeStepOut, bd.GenderStepOut)
					hcs := classHCs(bd.DatumOut, delta, bd.ClassStepOut, n)
					minDists := assignMinDistOutdoor(gender, age, maxDists)
					prestige := assignOutdoorPrestige(bowstyle, gender, age, maxDists)

					outdoorDict[groupname] = outdoorGroupData{
						classes:        classes,
						classesLong:    classesLong,
						classHC:        hcs,
						minDists:       minDists,
						maxDists:       maxDists,
						prestigeRounds: prestige,
					}
				}
			}
		}
	})
	return outdoorDict
}

// stdDists is the standard outdoor distance ladder in metres.
var stdDists = []float64{90, 70, 60, 50, 40, 30, 20, 15}

func assignMinDistOutdoor(gender Gender, age Age, maxDists []float64) []float64 {
	minMaxDist := minFloats(maxDists)
	maxDistIdx := distIndex(stdDists, minMaxDist)
	if maxDistIdx < 0 {
		maxDistIdx = 0
	}

	// U15 males and younger use one index pattern; everyone else uses another
	youngMale := age == Under15 || age == Under14 || age == Under12
	var idxs []int
	if gender == Male && !youngMale {
		idxs = []int{0, 0, 0, 0, 1, 2, 3, 4, 5}
	} else {
		idxs = []int{0, 0, 0, 0, 0, 1, 2, 3, 4}
	}

	return takeClip(stdDists, idxs, maxDistIdx)
}

func assignOutdoorPrestige(bowstyle Bowstyle, gender Gender, age Age, maxDists []float64) []string {
	prestigeImperial := []string{
		"york", "hereford", "bristol_i", "bristol_ii", "bristol_iii", "bristol_iv", "bristol_v",
	}
	prestigeMetric := []string{
		"wa1440_90", "wa1440_90_small", "wa1440_70", "wa1440_70_small",
		"wa1440_60", "wa1440_60_small",
		"metric_i", "metric_ii", "metric_iii", "metric_iv", "metric_v",
	}
	prestige720 := []string{
		"wa720_70", "wa720_60", "metric_122_50", "metric_122_40", "metric_122_30",
	}
	prestige720Compound := []string{
		"wa720_50_c", "metric_80_40", "metric_80_30",
	}
	prestige720Barebow := []string{
		"wa720_50_b", "metric_122_50", "metric_122_40", "metric_122_30",
	}

	var prestige []string
	var distCheck []string

	switch bowstyle {
	case Compound:
		prestige = append(prestige, prestige720Compound[0])
		distCheck = append(distCheck, prestige720Compound[1:]...)
	case Barebow:
		prestige = append(prestige, prestige720Barebow[0])
		distCheck = append(distCheck, prestige720Barebow[1:]...)
	default:
		prestige = append(prestige, prestige720[0])
		distCheck = append(distCheck, prestige720[1:]...)
		if gender == Male {
			if age == Over50 || age == Under18 {
				prestige = append(prestige, prestige720[1])
			} else if age == Under16 {
				prestige = append(prestige, prestige720[2])
			}
		}
	}

	distCheck = append(distCheck, prestigeImperial...)
	distCheck = append(distCheck, prestigeMetric...)

	minMaxDist := minFloats(maxDists)
	allRounds := allOutdoorRounds()
	for _, roundname := range distCheck {
		if r, ok := allRounds[roundname]; ok {
			if r.MaxDistance().Value >= minMaxDist {
				prestige = append(prestige, roundname)
			}
		}
	}

	return prestige
}

// CoaxOutdoorGroup maps non-outdoor bowstyles to their outdoor equivalent.
func CoaxOutdoorGroup(bowstyle Bowstyle, gender Gender, age Age) Category {
	coaxedBowstyle := bowstyle
	switch bowstyle {
	case Flatbow, Traditional:
		coaxedBowstyle = Barebow
	case CompoundLimited, CompoundBarebow:
		coaxedBowstyle = Compound
	}
	return Category{Bowstyle: coaxedBowstyle, Gender: gender, AgeGroup: age}
}

func isOutdoorBowstyle(b Bowstyle) bool {
	for _, ob := range OutdoorBowstyles {
		if b == ob {
			return true
		}
	}
	return false
}

func isValidGender(g Gender) bool {
	return g == Male || g == Female
}

func isValidAge(a Age) bool {
	for _, age := range AllAges {
		if a == age {
			return true
		}
	}
	return false
}

func getOutdoorGroupData(bowstyle Bowstyle, gender Gender, age Age) (outdoorGroupData, error) {
	if !isOutdoorBowstyle(bowstyle) {
		return outdoorGroupData{}, fmt.Errorf(
			"%s is not a recognised bowstyle for outdoor classifications", bowstyle)
	}
	if !isValidGender(gender) {
		return outdoorGroupData{}, fmt.Errorf(
			"%s is not a recognised gender for outdoor classifications", gender)
	}
	if !isValidAge(age) {
		return outdoorGroupData{}, fmt.Errorf(
			"%s is not a recognised age group for outdoor classifications", age)
	}
	dict := buildOutdoorDict()
	groupname := GroupName(bowstyle, gender, age)
	if gd, ok := dict[groupname]; ok {
		return gd, nil
	}
	return outdoorGroupData{}, fmt.Errorf("no outdoor data for group %q", groupname)
}

// CalculateOutdoorClassification returns the classification string for a score.
func CalculateOutdoorClassification(
	score float64,
	archeryRound *rounds.Round,
	bowstyle Bowstyle,
	gender Gender,
	age Age,
	strictRounds, strictDistance bool,
) (string, error) {
	if score < 0 || score > archeryRound.MaxScore() {
		return "", fmt.Errorf(
			"Invalid score of %.0f for a %s. Should be in range 0-%.0f.",
			score, archeryRound.Name, archeryRound.MaxScore(),
		)
	}

	scores, err := OutdoorClassificationScores(archeryRound, bowstyle, gender, age, strictRounds, strictDistance)
	if err != nil {
		return "", err
	}

	gd, err := getOutdoorGroupData(bowstyle, gender, age)
	if err != nil {
		return "", err
	}

	for i, classname := range gd.classes {
		classscore := float64(scores[i])
		if classscore < 0 || classscore > score {
			continue
		}
		return classname, nil
	}
	return "UC", nil
}

// OutdoorClassificationScores returns score thresholds for each class (descending order).
// -9999 indicates the class is not achievable on this round.
func OutdoorClassificationScores(
	archeryRound *rounds.Round,
	bowstyle Bowstyle,
	gender Gender,
	age Age,
	strictRounds, strictDistance bool,
) ([]int, error) {
	gd, err := getOutdoorGroupData(bowstyle, gender, age)
	if err != nil {
		return nil, err
	}

	allRounds := allOutdoorRounds()

	// Resolve round and codename
	roundname := archeryRound.Codename
	if strictRounds {
		if _, ok := allRounds[roundname]; !ok {
			return nil, fmt.Errorf(
				"This round is not recognised for outdoor classification. "+
					"Please select an appropriate option. (codename=%q)", roundname)
		}
		// Strip spots variant to get base round
		baseCodename := StripSpots(roundname)
		if base, ok := allRounds[baseCodename]; ok {
			archeryRound = base
		}
	}

	hcScheme := handicaps.MustScheme("AGB")

	n := len(gd.classes)
	classScores := make([]int, n)
	for i := range n {
		s := handicaps.ScoreForRound(hcScheme, gd.classHC[i], archeryRound, 0, true)
		classScores[i] = int(s)
	}

	// Prestige round check: non-prestige rounds cannot earn MB or above (classes 0,1,2)
	if strictRounds {
		isPrestige := false
		for _, pr := range gd.prestigeRounds {
			if pr == roundname {
				isPrestige = true
				break
			}
		}
		if !isPrestige {
			classScores[0] = -9999
			classScores[1] = -9999
			classScores[2] = -9999
		}
	}

	// Distance-based eligibility
	if strictDistance {
		isPrestige := false
		if strictRounds {
			for _, pr := range gd.prestigeRounds {
				if pr == roundname {
					isPrestige = true
					break
				}
			}
		}
		if !(strictRounds && isPrestige) {
			roundMaxDist := archeryRound.MaxDistance().Value
			for i := range classScores {
				if gd.minDists[i] > roundMaxDist {
					classScores[i] = -9999
				}
			}
		}
	}

	return classScores, nil
}

File: ts/src/classifications/outdoor.ts
import { newScheme, scoreForRound } from '../handicaps.js';
import { waOutdoorRounds, agbOutdoorImperialRounds, agbOutdoorMetricRounds, type Round, roundMaxScore } from '../rounds.js';
import {
  type Gender, type Age, type Bowstyle, type Category,
  AllGenders, AllAges, OutdoorBowstyles,
  Gender as G, Age as A, Bowstyle as B,
} from './data.js';
import {
  loadAgeData, loadBowstyleData, loadClassesOut, maxDistsForGender,
} from './jsonio.js';
import {
  groupName, getAgeGenderStep, classHCs, stripSpots,
  minFloats, distIndex, takeClip, fixRepeatedScores,
} from './utils.js';

interface OutdoorGroupData {
  classes: string[];
  classesLong: string[];
  classHC: number[];
  minDists: number[];
  maxDists: number[];
  prestigeRounds: string[];
}

let outdoorDict: Map<string, OutdoorGroupData> | null = null;

// Standard outdoor distance ladder (metres)
const STD_DISTS = [90, 70, 60, 50, 40, 30, 20, 15];

function allOutdoorRounds(): Map<string, Round> {
  const m = new Map<string, Round>();
  for (const [k, v] of agbOutdoorImperialRounds()) m.set(k, v);
  for (const [k, v] of agbOutdoorMetricRounds()) m.set(k, v);
  for (const [k, v] of waOutdoorRounds()) m.set(k, v);
  return m;
}

function assignMinDistOutdoor(gender: Gender, age: Age, maxDists: number[]): number[] {
  const minMaxDist = minFloats(maxDists);
  let maxDistIdx = distIndex(STD_DISTS, minMaxDist);
  if (maxDistIdx < 0) maxDistIdx = 0;

  const youngMale = age === A.Under15 || age === A.Under14 || age === A.Under12;
  const idxs = (gender !== G.Female && !youngMale)
    ? [0, 0, 0, 0, 1, 2, 3, 4, 5]
    : [0, 0, 0, 0, 0, 1, 2, 3, 4];
  return takeClip(STD_DISTS, idxs, maxDistIdx);
}

function assignOutdoorPrestige(bowstyle: Bowstyle, gender: Gender, age: Age, maxDists: number[]): string[] {
  const prestigeImperial = [
    'york', 'hereford', 'bristol_i', 'bristol_ii', 'bristol_iii', 'bristol_iv', 'bristol_v',
  ];
  const prestigeMetric = [
    'wa1440_90', 'wa1440_90_small', 'wa1440_70', 'wa1440_70_small',
    'wa1440_60', 'wa1440_60_small',
    'metric_i', 'metric_ii', 'metric_iii', 'metric_iv', 'metric_v',
  ];
  const prestige720 = [
    'wa720_70', 'wa720_60', 'metric_122_50', 'wa720_40', 'metric_122_40', 'metric_122_30',
  ];
  const prestige720Compound = [
    'wa720_50_c', 'metric_80_50', 'wa720_40_c', 'metric_80_40', 'metric_80_30',
  ];
  const prestige720Barebow = [
    'wa720_50_b', 'metric_122_50', 'wa720_40', 'metric_122_40', 'wa720_30_b', 'metric_122_30',
  ];

  const youngJunior = age === A.Under15 || age === A.Under14 || age === A.Under12;
  let prestige: string[] = [];
  let distCheck: string[] = [];

  if (bowstyle === B.Compound) {
    prestige.push(...prestige720Compound.slice(0, 2));
    distCheck.push(...prestige720Compound.slice(2));
    if (youngJunior) prestige.push(...prestige720Compound.slice(2, 4));
  } else if (bowstyle === B.Barebow) {
    prestige.push(prestige720Barebow[0]);
    distCheck.push(...prestige720Barebow.slice(1));
    if (youngJunior) prestige.push(...prestige720Barebow.slice(2));
  } else {
    prestige.push(prestige720[0]);
    distCheck.push(...prestige720.slice(1));
    if (youngJunior) prestige.push(...prestige720.slice(3, 5));
    if (gender === G.Open || gender === G.Male) {
      if (age === A.Over50 || age === A.Under18) {
        prestige.push(prestige720[1]);
      } else if (age === A.Under16) {
        prestige.push(prestige720[2]);
      }
    }
  }

  distCheck.push(...prestigeImperial, ...prestigeMetric);
  const minMaxDist = minFloats(maxDists);
  const allRounds = allOutdoorRounds();
  for (const rname of distCheck) {
    const r = allRounds.get(rname);
    if (r) {
      const md = r.passes.reduce((m, p) => Math.max(m, p.target.distance), 0);
      if (md >= minMaxDist) prestige.push(rname);
    }
  }
  return prestige;
}

function buildOutdoorDict(): Map<string, OutdoorGroupData> {
  if (outdoorDict !== null) return outdoorDict;
  const agbAgeData = loadAgeData();
  const agbBowstyleData = loadBowstyleData();
  const classInfo = loadClassesOut();
  const { classes, classes_long: classesLong } = classInfo;
  const n = classes.length;

  outdoorDict = new Map();
  for (const bowstyle of OutdoorBowstyles) {
    for (const gender of AllGenders) {
      for (const age of AllAges) {
        const gname = groupName(bowstyle, gender, age);
        const ad = agbAgeData[age];
        const bd = agbBowstyleData[bowstyle];
        const maxDists = maxDistsForGender(ad, gender);
        const ageCat = ad.step;
        const delta = getAgeGenderStep(gender, ageCat, bd.ageStep_out, bd.genderStep_out);
        const hcs = classHCs(bd.datum_out, delta, bd.classStep_out, n);
        const minDists = assignMinDistOutdoor(gender, age, maxDists);
        const prestige = assignOutdoorPrestige(bowstyle, gender, age, maxDists);
        outdoorDict.set(gname, { classes, classesLong, classHC: hcs, minDists, maxDists, prestigeRounds: prestige });
      }
    }
  }
  return outdoorDict;
}

function isOutdoorBowstyle(b: Bowstyle): boolean { return OutdoorBowstyles.includes(b); }
function isValidGender(g: Gender): boolean { return g === G.Male || g === G.Female || g === G.Open; }
function isValidAge(a: Age): boolean { return AllAges.includes(a); }

function getOutdoorGroupData(bowstyle: Bowstyle, gender: Gender, age: Age): OutdoorGroupData {
  if (!isOutdoorBowstyle(bowstyle))
    throw new Error(`${bowstyle} is not a recognised bowstyle for outdoor classifications`);
  if (!isValidGender(gender))
    throw new Error(`${gender} is not a recognised gender for outdoor classifications`);
  if (!isValidAge(age))
    throw new Error(`${age} is not a recognised age group for outdoor classifications`);
  const gd = buildOutdoorDict().get(groupName(bowstyle, gender, age));
  if (!gd) throw new Error(`No outdoor data for group "${groupName(bowstyle, gender, age)}"`);
  return gd;
}

export function coaxOutdoorGroup(bowstyle: Bowstyle, gender: Gender, age: Age): Category {
  let coaxed = bowstyle;
  if (bowstyle === B.Flatbow || bowstyle === B.Traditional) coaxed = B.Barebow;
  else if (bowstyle === B.CompoundLimited || bowstyle === B.CompoundBarebow) coaxed = B.Compound;
  return { bowstyle: coaxed, gender, ageGroup: age };
}

export function outdoorClassificationScores(
  archeryRound: Round,
  bowstyle: Bowstyle,
  gender: Gender,
  age: Age,
  strictRounds: boolean,
  strictDistance: boolean,
): number[] {
  const gd = getOutdoorGroupData(bowstyle, gender, age);
  const allRounds = allOutdoorRounds();
  let round = archeryRound;
  let roundname = archeryRound.codename;

  if (strictRounds) {
    if (!allRounds.has(roundname)) {
      throw new Error(
        `This round is not recognised for outdoor classification. Please select an appropriate option. (codename="${roundname}")`
      );
    }
    const base = stripSpots(roundname);
    const baseRound = allRounds.get(base);
    if (baseRound) round = baseRound;
  }

  const hcScheme = newScheme('AGB');
  const classScores = gd.classHC.map(hc => Math.trunc(scoreForRound(hcScheme, hc, round, 0, true)));

  // Gap/max-score handling
  const maxScore = roundMaxScore(round);
  for (let i = 0; i < gd.classHC.length; i++) {
    const nextScore = Math.trunc(scoreForRound(hcScheme, Math.floor(gd.classHC[i]) + 1, round, 0, true));
    if (nextScore === classScores[i]) {
      classScores[i] = classScores[i] === Math.trunc(maxScore) ? -9999 : classScores[i] + 1;
    }
  }
  fixRepeatedScores(classScores, maxScore);

  if (strictRounds) {
    const isPrestige = gd.prestigeRounds.includes(roundname);
    if (!isPrestige) { classScores[0] = -9999; classScores[1] = -9999; classScores[2] = -9999; }
  }

  if (strictDistance) {
    const isPrestige = strictRounds && gd.prestigeRounds.includes(roundname);
    if (!isPrestige) {
      const roundMaxDist = round.passes.reduce((m, p) => Math.max(m, p.target.distance), 0);
      for (let i = 0; i < classScores.length; i++) {
        if (gd.minDists[i] > roundMaxDist) classScores[i] = -9999;
      }
    }
  }

  return classScores;
}

export function calculateOutdoorClassification(
  score: number,
  archeryRound: Round,
  bowstyle: Bowstyle,
  gender: Gender,
  age: Age,
  strictRounds: boolean,
  strictDistance: boolean,
): string {
  const maxScore = roundMaxScore(archeryRound);
  if (score < 0 || score > maxScore) {
    throw new Error(
      `Invalid score of ${Math.trunc(score)} for a ${archeryRound.name}. Should be in range 0-${Math.trunc(maxScore)}.`
    );
  }
  const scores = outdoorClassificationScores(archeryRound, bowstyle, gender, age, strictRounds, strictDistance);
  const gd = getOutdoorGroupData(bowstyle, gender, age);
  for (let i = 0; i < gd.classes.length; i++) {
    if (scores[i] < 0 || scores[i] > score) continue;
    return gd.classes[i];
  }
  return 'UC';
}
